src/Microsoft.Framework.TestHost/ProjectCommand.cs

```diff
@@ -25,12 +32,105 @@ public class ProjectCommand
             var newServices = new ServiceProvider(services);
             newServices.Add(typeof(IApplicationEnvironment), environment);
 
-            var applicationHost = new Microsoft.Framework.ApplicationHost.Program(
-                (IAssemblyLoaderContainer)services.GetService(typeof(IAssemblyLoaderContainer)),
-                environment,
-                newServices);
+            var newHostOptions = new DefaultHostOptions()
+            {
+                ApplicationBaseDirectory = project.ProjectDirectory,
+                CompilationServerPort = oldHostOptions.CompilationServerPort,
+                Configuration = oldHostOptions.Configuration,
+                PackageDirectory = oldHostOptions.PackageDirectory,
+                TargetFramework = oldHostOptions.TargetFramework,
+                WatchFiles = oldHostOptions.WatchFiles,
+            };
 
-            return await applicationHost.Main(args);
+            var host = new DefaultHost(newHostOptions, newServices);
+
+            var commandText = host.Project.Commands[command];
+            var replacementArgs = CommandGrammar.Process(
+                commandText, 
+                (key) => GetVariable(environment, key))
+                .ToArray();
+
+            newHostOptions.ApplicationName = replacementArgs[0];
+            args = replacementArgs.Skip(1).Concat(args).ToArray();
+
+            if (string.IsNullOrEmpty(newHostOptions.ApplicationName) ||
+                string.Equals(newHostOptions.ApplicationName, "run", StringComparison.Ordinal))
+            {
+                newHostOptions.ApplicationName = host.Project.EntryPoint ?? host.Project.Name;
+            }
+
+            return await ExecuteMain(host, newHostOptions.ApplicationName, args);
+        }
+
+        private static string GetVariable(IApplicationEnvironment environment, string key)
+        {
+            if (string.Equals(key, "env:ApplicationBasePath", StringComparison.OrdinalIgnoreCase))
+            {
+                return environment.ApplicationBasePath;
+            }
+            if (string.Equals(key, "env:ApplicationName", StringComparison.OrdinalIgnoreCase))
+            {
+                return environment.ApplicationName;
+            }
+            if (string.Equals(key, "env:Version", StringComparison.OrdinalIgnoreCase))
+            {
+                return environment.Version;
+            }
+            if (string.Equals(key, "env:TargetFramework", StringComparison.OrdinalIgnoreCase))
+            {
+                return environment.RuntimeFramework.Identifier;
+            }
+
+            return Environment.GetEnvironmentVariable(key);
+        }
```

Comment:
This jazz is still needed because I need to process the command in the project.json into something callable.

If you're so inclined we could just put this logic into default host.